Some elaboration of how to read compiler errors. (#104)
bhoffman0/CSAwesome: _sources/Unit1-Getting-Started/topic-1-2-java-intro.rst

> |CodingEx| **Coding Exercise: Compile Time Error 1**

Reviewer: This got left behind. I think it goes above the exercise on line 343.

> Reading error messages

Reviewer: ```suggestion
Reading Error Messages
```

----------------------

> In the exercise below you should click on the |runbutton| button below to

Reviewer: ```suggestion
In the exercise below, you should click on the |runbutton| button to
```

the code that you need to fix. Before you try to fix it, look for the error
message below the code. This is called a **compile time error** because it is an
error detected by the compiler.

It will be a lot easier to find the problem, and other problems like it, if you
understand how to read the error message. So let’s take a closer look. It will
look something like this:

.. code-block:: text

   FirstClass.java:5: error: unclosed string literal
          System.out.println("Hi there!);
                             ^
   1 error

The first line starts with the name of the file that was being compiled.
Because you submitted your code via this web page you didn't actually name
the file but it was named for you to match the name of your class. Then
there's a colon (``:``) followed by a number. That number tells you the line
number in the file where the compiler detected the error, in this case
line 5.

Error messages aren't always 100% accurate about where the error actually is;
sometimes you actually need to change something a bit earlier in the program
and sometimes a bit later. But the line number is the best place to start
looking.

> After the line number and another colon is the actual error message. These

Reviewer: ```suggestion
After the line number and another colon, you will find the actual error message. These
```

vocabulary they will become more meaningful but they almost always contain
some useful clues. For instance take this error message: “unclosed string
literal”. You may not know what a string literal is (yet) but “unclosed”
suggests something was opened and then not closed. Keep that thought in mind.

Now look at the next two lines. The very next line is just the line of code
from your program. But below that is a very important line containing a
single caret (``^``) positioned to point at exactly where in the line the
Java compiler thinks the problem is. In this case it’s pointing at the
quotation mark (``”``) before “Hi”. So it’s complaining about something being
unclosed and it’s pointing us at a quotation mark. Usually quotation marks
come in pairs called the open quote and the close quote, right? And if you
look at the line of code you’ll see that there’s no closing quotation mark.
Could that be the problem? Try adding a quotation mark and see if that fixes
it!

It is worth getting in the habit of really reading error messages. The people
who wrote the Java compiler put in lot of work to try to make the error

> messages useful. If you actually read them you’ll soon learn to recognize

Reviewer: ```suggestion
messages useful. If you read them, you’ll soon learn to recognize
```

|CodingEx| **Coding Exercise: Compile Time Error 3**

Click on the |runbutton| button below to try and run the following code. What is
wrong this time? The first error message will probably point you directly at the
problem. But after you fix that you’ll probably get a `new` error! That happens
sometimes because the compiler can’t even detect the second error until the
first one is fixed.

This is another reason it’s important to read the error message; if you just

> start trying to fix things without actually reading the error message you might

Reviewer: Maybe too snarky? :-)

Author: Is it better now?
